orchestration/scripts/query-non-null-rows.sh

```diff
@@ -0,0 +1,39 @@
+set -euo pipefail
+
+# Copy-pasted from StackOverflow.
+function join_by { local d=$1; shift; echo -n "$1"; shift; printf "%s" "${@/#/$d}"; }
+
+# Point to BQ metadata we expect to be present on disk.
+declare -r TABLE_DIR=/bq-metadata/${TABLE}
+declare -r PK_COLS=$(cat ${TABLE_DIR}/primary-keys)
+
+# Build the WHERE clause of the SQL query.
+declare -a COMPARISONS=()
+for c in ${PK_COLS//,/ }; do
+  COMPARISONS+=("${c} IS NOT NULL")
+done
+declare -r FULL_DIFF=$(join_by ' AND ' "${COMPARISONS[@]}")
+
+declare -r TARGET_TABLE=${TABLE}_values_${OUTPUT_SUFFIX}
+
+# Pull everything but the row ID from rows with non-null primary keys.
```

Comment:
I am a bit confused about why we're expecting primary keys to sometimes be null, since primary keys are usually required to be non-null. Are these not the primary keys of the table that's being queried?

Reply:
We're doing a full outer join between the staged data and the data already in Jade, and selecting out:
1. Just the row ID column from Jade, and
2. All of the data columns from the staging table

Because we filter out results where all the data columns are identical, we end up with 3 situations:
1. Row ID column is null, primary-key columns are not null: A row describing a brand-new entity is present in the staging data
2. Row ID column is not null, primary-key columns are not null: A row describing a previously-ingested entity is present in the staging data
3. Row ID column is not null, primary-key columns are null: A previously-ingested entity has no row in the staging data

Reply:
So, we don't expect the PKs to ever be null within the data, but it's expected that some nulls will pop out on the full outer join.